candle-transformers/src/models/metavoice.rs

```diff
@@ -0,0 +1,316 @@
+#[allow(clippy::upper_case_acronyms)]
+enum MLP {
+    Gelu {
+        c_fc: Linear,
+        c_proj: Linear,
+    },
+    Swiglu {
+        w1: Linear,
+        w3: Linear,
+        c_proj: Linear,
+    },
+}
+
+impl MLP {
+    fn new(cfg: &GPTConfig, vb: VarBuilder) -> Result<Self> {
+        let hidden_dim = 4 * cfg.n_embd;
+        let slf = match cfg.nonlinearity_type {
+            NonLinearityType::Gelu => {
+                let c_fc = linear(cfg.n_embd, hidden_dim, cfg.bias, vb.pp("c_fc"))?;
+                let c_proj = linear(hidden_dim, cfg.n_embd, cfg.bias, vb.pp("c_proj"))?;
+                Self::Gelu { c_fc, c_proj }
+            }
+            NonLinearityType::Swiglu => {
+                let hidden_dim = (2 * hidden_dim) / 3;
+                let swiglu_multiple_of = match cfg.swiglu_multiple_of {
+                    None => candle::bail!("swiglu-multiple-of has to be set"),
+                    Some(smo) => smo,
+                };
+                let hidden_dim =
+                    swiglu_multiple_of * (hidden_dim + swiglu_multiple_of - 1) / swiglu_multiple_of;
+                let w1 = linear(cfg.n_embd, hidden_dim, cfg.bias, vb.pp("w1"))?;
+                let w3 = linear(cfg.n_embd, hidden_dim, cfg.bias, vb.pp("w3"))?;
+                let c_proj = linear(hidden_dim, cfg.n_embd, cfg.bias, vb.pp("c_proj"))?;
+                Self::Swiglu { w1, w3, c_proj }
+            }
+        };
+        Ok(slf)
+    }
+}
+
+impl Module for MLP {
+    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
+        match self {
+            Self::Gelu { c_fc, c_proj } => xs.apply(c_fc)?.gelu()?.apply(c_proj),
+            Self::Swiglu { w1, w3, c_proj } => {
+                let w1 = xs.apply(w1)?;
+                let w3 = xs.apply(w3)?;
+                (w1.silu()? * w3)?.apply(c_proj)
+            }
+        }
+    }
+}
+
+// https://github.com/metavoiceio/metavoice-src/blob/11550bb4e8a1ad032cc1556cc924f7a4e767cbfa/fam/llm/layers/combined.py#L7
+struct Block {
+    ln_1: Norm,
+    ln_2: Norm,
+    attn: SelfAttention,
+    mlp: MLP,
+}
+
+impl Block {
+    fn new(cfg: &GPTConfig, vb: VarBuilder) -> Result<Self> {
+        let ln_1 = Norm::new(cfg, vb.pp("ln_1"))?;
+        let ln_2 = Norm::new(cfg, vb.pp("ln_2"))?;
+        let attn = SelfAttention::new(cfg, vb.pp("attn"))?;
+        let mlp = MLP::new(cfg, vb.pp("mlp"))?;
+        Ok(Block {
+            ln_1,
+            ln_2,
+            attn,
+            mlp,
+        })
+    }
+}
+
+impl Module for Block {
+    fn forward(&self, xs: &Tensor) -> Result<Tensor> {
+        let xs = (xs + xs.apply(&self.ln_1)?.apply(&self.attn))?;
+        let xs = (&xs + xs.apply(&self.ln_2)?.apply(&self.mlp))?;
+        Ok(xs)
+    }
+}
+
+// https://github.com/metavoiceio/metavoice-src/blob/11550bb4e8a1ad032cc1556cc924f7a4e767cbfa/fam/llm/model.py#L79
+#[allow(unused)]
+#[allow(clippy::upper_case_acronyms)]
+pub struct GPT {
+    wtes: Vec<candle_nn::Embedding>,
```

Comment:
nit: this is an artefact of some legacy code on our part, we're doing single hierarchy flattened interleaved output, a single Embedding layer should be fine